src/openloco/localisation/stringmgr.cpp

```diff
@@ -1,23 +1,337 @@
+
+                // lea     args, towns[args]
+                void* town_name = (void*) &town->name;
+
+                // call    format_string
+                format_string(buffer, id, town_name);
+
+                return;
+            }
+            else if (id == 41024)
+            {
+                auto temp = args;
+
+                // movzx   ecx, word ptr [ecx]
+                uint16_t town_id = *(uint16_t*) args;
+                args = (uint8_t*) args + 2;
+
+                // movzx   id, word ptr towns[args]
+                auto town = townmgr::get(town_id);
+
+                // call    format_string
+                format_string(buffer, town->name, nullptr);
+
+                // pop     args
+                args = temp;
+
+                return;
+            }
+            else
+            {
+                throw std::out_of_range("format_string: invalid string id: " + id);
+            }
+        }
+        else
+        {
+            registers regs;
+            regs.eax = id;
+            regs.edi = (uint32_t)buffer;
+            regs.ecx = (uint32_t)args;
+
+            char* sourceStr = get_string(id);
+
+            while (uint8_t ch = *sourceStr++)
+            {
+                if (ch <= 0x1F)
+                {
+                    if (ch == 0)
+                    {
+                        *buffer = '\0';
+                        return;
+                    }
+                    else if (ch <= 4)
+                    {
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                    }
+                    else if (ch <= 16)
+                    {
+                        *buffer++ = ch;
+                    }
+                    else if (ch <= 22)
+                    {
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                    }
+                    else
+                    {
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                        ch = *sourceStr++
+                        *buffer++ = ch;
+                    }
+                }
+                else if (ch < '}') // 0x7B
+                {
+                    *buffer++ = ch;
+                }
+                else if (ch < 0x90)
+                {
+                    switch (ch)
+                    {
+                        case 123 + 0:
+                        {
+                            uint32_t value = *args;
+                            args = (uint8_t*) args + 4;
+                            format_comma(value, buffer);
```

Comment:
These format functions should probably affect the position in `buffer`

Reply:
Right. We could make them return the register value used for `buffer` (`edi`) back for assignment?

Reply:
That seems like a sensible approach 👍